pkg/proto/eth_transaction.go

```diff
@@ -458,6 +477,98 @@ func (tx *EthereumTransaction) RawSignatureValues() (v, r, s *big.Int) {
 	return tx.inner.rawSignatureValues()
 }
 
+func validateEthereumTx(tx *EthereumTransaction) error {
+	switch tx.TxKind.(type) {
+	case *EthereumTransferWavesTxKind:
+		res := new(big.Int).Div(tx.inner.value(), big.NewInt(int64(DiffEthWaves)))
+		if ok := res.IsInt64(); !ok {
+			return errors.Errorf("failed to convert amount from ethreum transaction (big int) to int64. value is %d", tx.Value())
+		}
+		tx.value = res.Int64()
+		return nil
+	case *EthereumTransferAssetsErc20TxKind:
+		return nil
+	case *EthereumInvokeScriptTxKind:
+		return nil
+	default:
+		return errors.New("failed to check ethereum transaction, wrong kind of tx")
+	}
+}
+
+func GuessEthereumTransactionKind(data []byte) (int64, error) {
+	if len(data) == 0 {
+		return EthereumTransferWavesKind, nil
+	}
+
+	selectorBytes := data
+	if len(data) < ethabi.SelectorSize {
+		return 0, errors.Errorf("length of data from ethereum transaction is less than %d", ethabi.SelectorSize)
+	}
+	selector, err := ethabi.NewSelectorFromBytes(selectorBytes[:ethabi.SelectorSize])
+	if err != nil {
+		return 0, errors.Wrap(err, "failed to guess ethereum transaction kind")
+	}
+
+	if ethabi.IsERC20TransferSelector(selector) {
+		return EthereumTransferAssetsKind, nil
+	}
+
+	return EthereumInvokeKind, nil
+}
+
+// note: in this method not all of the fields are filled
+func GetEthereumTransactionKind(ethTx EthereumTransaction) (EthereumTransactionKind, error) {
+	txKind, err := GuessEthereumTransactionKind(ethTx.Data())
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to guess ethereum tx kind")
+	}
+
+	switch txKind {
+	case EthereumTransferWavesKind:
+		return NewEthereumTransferWavesTxKind(), nil
+	case EthereumTransferAssetsKind:
+		db := ethabi.NewErc20MethodsMap()
+		decodedData, err := db.ParseCallData(ethTx.Data())
+		if err != nil {
+			return nil, errors.Errorf("failed to parse ethereum data")
+		}
+		if len(decodedData.Inputs) != ethabi.NumberOfERC20TransferArguments {
```

Comment:
This check is unnecessary. `ethabi.GetERC20TransferArguments` does the same check inside.